// !LANGUAGE: +NewInference
// !DIAGNOSTICS: -UNUSED_PARAMETER -UNUSED_VARIABLE

fun <T> bar(i: T): T = i
fun foo(i: Int) = i

fun dontRun(body: () -> Unit) = Unit

class Case1 {
    fun test() {
        dontRun { <!UNREACHABLE_CODE!>val x = <!IMPLICIT_NOTHING_AS_TYPE_PARAMETER!>bar<!>(<!><!IMPLICIT_NOTHING_AS_TYPE_PARAMETER, IMPLICIT_NOTHING_AS_TYPE_PARAMETER!>bar<!> <!TYPE_MISMATCH, TYPE_MISMATCH, TYPE_MISMATCH!>{ -> <!IMPLICIT_NOTHING_AS_TYPE_PARAMETER, IMPLICIT_NOTHING_AS_TYPE_PARAMETER!>bar<!> <!TYPE_MISMATCH, TYPE_MISMATCH, TYPE_MISMATCH!>{ -> 2}<!> }<!><!UNREACHABLE_CODE!>)<!> }
    }
}